### PR TITLE
initial implementation of bufferline

### PR DESCRIPTION
**Pull request by aaron404**
- adds new "bufferline" setting, which enables rendering of a line displaying currently open buffers
- has 3 modes `always` (always show), `never` (never show, default), `multiple` (show if num buffers > 1)

helix2

[#497](https://github.com/helix-editor/helix/issues/497)

### CODE REVIEW

**Comment by the-mikedavis** [comment]
I think we should call this bufferline instead of tabs. Tabs as in https://github.com/helix-editor/helix/issues/2295 behave differently.

**Review by the-mikedavis — COMMENTED** [review]
Could you add some docs on this to the book (https://github.com/helix-editor/helix/blob/master/book/src/configuration.md)?

**Comment by lazytanuki** [comment]
It could be nice to have an option that, similarly to some other editors, only shows the bufferline when there is more than one buffer to display in it.

File: helix-term/src/ui/editor.rs [hunk 1]
```diff
@@ -541,6 +554,43 @@ impl EditorView {
...
+                .file_name()
+                .unwrap_or_default()
+                .to_str()
+                .unwrap();
+
+            let style = if view.doc == doc.id() {
+                theme.get("ui.background")
+            } else {
+                theme.get("ui.statusline")
+            };
```
Comment (sudormrfbin) [inline helix-term/src/ui/editor.rs:581]: Introducing new theme scopes like `ui.bufferline` and `ui.bufferline.active` which fall back to these values would be a nice addition.

**Review by sudormrfbin — COMMENTED** [review]
A couple of improvements that could be implemented (not necessarily in this PR, since I'd say it's at the sweet spot for ease of reviewing):

- Showing previous path components when two buffers have files with the same name.
- Taking the width of the viewport into account and aligning the tabs so that current buffer name is visible even on small terminals.

File: helix-term/src/ui/editor.rs (hunk 1 above)
Comment (aaron404) [reply helix-term/src/ui/editor.rs:581]: Is there a straighforward way to have "ui.bufferline" fallback to "ui.statusline"? Reading the docs it seems that it would simply fallback to "ui".

File: book/src/configuration.md [hunk 2]
```diff
@@ -45,6 +45,7 @@ hidden = false
 | `auto-info` | Whether to display infoboxes | `true` |
 | `true-color` | Set to `true` to override automatic detection of terminal truecolor support in the event of a false negative. | `false` |
 | `rulers` | List of column positions at which to display the rulers. Can be overridden by language specific `rulers` in `languages.toml` file. | `[]` |
+| `bufferline` | Renders a line at the top of the editor displaying open buffers. Can be `always`, `never` or `multiple` (only shown if more than one buffer is in use) | `multiple`
```
Comment (zen3ger) [inline book/src/configuration.md:48]: not sure how forgiving the markdown renderer is, but there's a missing `|` at the end.

File: helix-term/src/ui/editor.rs [hunk 3]
```diff
@@ -541,6 +541,44 @@ impl EditorView {
...
+            } else {
+                editor.theme.get("ui.statusline")
+            };
+
+            surface.set_string(
+                1 + viewport.x + len as u16,
+                viewport.y,
+                format!(" {} {}", fname, if doc.is_modified() { "● " } else { "" }),
+                style,
+            );
```
Comment (the-mikedavis) [inline helix-term/src/ui/editor.rs:569]: it looks like this block is redundant because of the call to `surface.set_string` below

File: helix-term/src/ui/editor.rs [hunk 4]
```diff
@@ -541,6 +541,44 @@ impl EditorView {
...
+
+            let (text, offset) = if doc.is_modified() {
+                (format!(" {} ● ", fname), fname.len() + 4)
```
Comment (the-mikedavis) [inline helix-term/src/ui/editor.rs:572]:
```suggestion
                (format!(" {}[+] ", fname), fname.len() + 5)
```

Let's reuse the modified indicator from the status-line for consistency

File: helix-term/src/ui/editor.rs (hunk 1 above)
Comment (the-mikedavis) [reply helix-term/src/ui/editor.rs]: You could do something like `theme.try_get("ui.bufferline").unwrap_or_else(|| theme.get("ui.statusline")`. That would behave differently from how other theme fallbacks work though, so I think we should just introduce a new `ui.bufferline` and copy `ui.statusline` values in the existing themes

**Comment by aaron404** [comment]
> Though, I'm not sure if quick switching and adding shortcuts should be within the scope of this initial PR.

Yeah I'd be happy to do this as a follow up PR if there was interest in it, but for now I think it is out of scope

**Comment by lazytanuki** [comment]
Another thing that would be nice is to make it automatically scroll to display the currently viewed buffer. At the moment, if we have too many open buffers, the bufferline will overflow at the end. A bit like this from nvim-cokeline:

this

However, implementing the "natural" behavior would require tracking the bufferline position, in order for the render to know where was the focus during the previous render.

An alternative to postpone this could be to shorten the filenames when the bufferline gets too wide.

File: book/src/configuration.md [hunk 5]
```diff
@@ -45,6 +45,7 @@ hidden = false
 | `auto-info` | Whether to display infoboxes | `true` |
 | `true-color` | Set to `true` to override automatic detection of terminal truecolor support in the event of a false negative. | `false` |
 | `rulers` | List of column positions at which to display the rulers. Can be overridden by language specific `rulers` in `languages.toml` file. | `[]` |
+| `bufferline` | Renders a line at the top of the editor displaying open buffers. Can be `always`, `never` or `multiple` (only shown if more than one buffer is in use) | `multiple` |
```
Comment (the-mikedavis) [inline book/src/configuration.md:48]: I'd prefer this default to `never`. I think it falls under the category of features that some really like and others find distracting or overly verbose like visible whitespace (also indent guides for example).

File: helix-term/src/ui/editor.rs [hunk 6]
```diff
@@ -541,6 +541,45 @@ impl EditorView {
...
+                .unwrap();
+
+            let style = if view!(editor).doc == doc.id() {
+                // editor.theme.get("ui.background")
+                editor
+                    .theme
+                    .try_get("ui.bufferline.active")
+                    .unwrap_or_else(|| editor.theme.get("ui.background"))
+            } else {
+                // editor.theme.get("ui.statusline")
```
Comment (the-mikedavis) [inline helix-term/src/ui/editor.rs:565]: these comments can be cleaned up

**Comment by sbromberger** [comment]
I might be missing something, but how do I specify a separator between the open buffers? Right now it looks like

bufferline

**Comment by sbromberger** [comment]
I've been using this for the past several days and have come to rely on it. No glaring issues that I see. FWIW, I would love to see this merged soon.

File: book/src/configuration.md (hunk 5 above)
Comment (aaron404) [reply book/src/configuration.md:48]: There was some feedback earlier that `multiple` would be a good default, but if there is agreement I am happy to change it back to never.

**Comment by aaron404** [comment]
> I might be missing something, but how do I specify a separator between the open buffers?

I haven't added any way to configure this and hadn't planned on it. Perhaps someone could do a follow up to add this functionality.

**Comment by sbromberger** [comment]
> > I might be missing something, but how do I specify a separator between the open buffers?
> 
> I haven't added any way to configure this and hadn't planned on it. Perhaps someone could do a follow up to add this functionality.

Yes - sorry - I absolutely do NOT want to hold this PR up. I'll take a look at what needs to be implemented once this goes in. (I'll probably have to make some changes due to #2831 if that gets merged, too.)

File: book/src/configuration.md (hunk 5 above)
Comment (sudormrfbin) [reply book/src/configuration.md:48]: I would personally vote for having `multiple` by default since it's more discoverable that way and can be easily disabled in the config. But then again that argument could be extended to every single new feature so :)

Comment (aaron404) [reply book/src/configuration.md:48]: I am personally also of the opinion that `multiple` is the better default option. Anyone who cares that much about the wasted space of a single line will probably have a bunch of custom options set in their config anyways, and adding one more to hide it should be trivial.

File: helix-term/src/ui/editor.rs [hunk 7]
```diff
@@ -578,6 +578,43 @@ impl EditorView {
...
+                    .theme
+                    .try_get("ui.bufferline")
+                    .unwrap_or_else(|| editor.theme.get("ui.statusline"))
+            };
+
+            let (text, offset) = if doc.is_modified() {
+                (format!(" {}[+] ", fname), fname.len() + 5)
+            } else {
+                (format!(" {} ", fname), fname.len() + 2)
+            };
```
Comment (sudormrfbin) [inline helix-term/src/ui/editor.rs:611]:
```suggestion
            let text = format!(" {}{} ", fname, if doc.is_modified() { "[+]" } else { "" });
            let offset = text.len();
```

File: book/src/configuration.md (hunk 5 above)
Comment (the-mikedavis) [reply book/src/configuration.md:48]: I'm curious what @archseer thinks about the default behavior.

I was going to say that the default keymap favors the buffer picker for switching buffers but that's not actually true: `gn`/`gp` are the same number of keystrokes. I'm definitely biased as I only ever use the buffer picker.

> Anyone who cares that much about the wasted space of a single line will probably have a bunch of custom options set in their config anyways, and adding one more to hide it should be trivial.

It's more consistent with the current configuration options to make the reverse argument: anyone who wants to enable a cosmetic feature will probably have a bunch of custom options set in their config anyways, and adding one more to show it should be trivial. Examples: visible whitespace, indent guides, cursorline, rulers, custom cursor shapes.

> it's more discoverable [enabled by default]

Hmm yeah I agree. IMO though a toggle mode like #2849 would be the ideal approach to that - enabling new features by default as a general policy might end up with a busy/overwhelming default experience.

**Comment by archseer** [comment]
Bufferline is not something I'd like to enable by default, in my opinion it encourages the wrong workflow (treat views == tabs), it's just a less featureful version of space+b buffer list. But I understand people are used to sublime/vscode-like tabs.

File: helix-term/src/ui/editor.rs [hunk 8]
```diff
@@ -578,6 +578,40 @@ impl EditorView {
...
+        surface.clear_with(viewport, editor.theme.get("ui.statusline"));
+        let mut len = 0usize;
+        for doc in editor.documents() {
+            let fname = doc
+                .path()
+                .unwrap_or(&pb)
+                .file_name()
+                .unwrap_or_default()
+                .to_str()
+                .unwrap();
```
Comment (archseer) [inline helix-term/src/ui/editor.rs:593]: No unwraps, use unwrap_or_default()

File: helix-term/src/ui/editor.rs [hunk 9]
```diff
@@ -578,6 +578,40 @@ impl EditorView {
         }
     }
 
+    /// Render bufferline at the top
+    pub fn render_bufferline(editor: &Editor, viewport: Rect, surface: &mut Surface) {
+        let pb = PathBuf::from(SCRATCH_BUFFER_NAME); // default filename to use for scratch buffer
+        surface.clear_with(viewport, editor.theme.get("ui.statusline"));
```
Comment (lazytanuki) [inline helix-term/src/ui/editor.rs:584]: Could we have a `ui.bufferline.background` style here ?

```suggestion
        surface.clear_with(
            viewport,
            editor
                .theme
                .try_get("ui.bufferline.background")
                .unwrap_or_else(|| editor.theme.get("ui.statusline")),
        );
```

File: helix-term/src/ui/editor.rs [hunk 10]
```diff
@@ -578,6 +578,40 @@ impl EditorView {
         }
     }
 
+    /// Render bufferline at the top
+    pub fn render_bufferline(editor: &Editor, viewport: Rect, surface: &mut Surface) {
+        let pb = PathBuf::from(SCRATCH_BUFFER_NAME); // default filename to use for scratch buffer
```
Comment (the-mikedavis) [inline helix-term/src/ui/editor.rs:583]: this is a bit of a nit but this variable name could be more descriptive: `scratch` or maybe `scratch_path`

File: helix-term/src/ui/editor.rs [hunk 11]
```diff
@@ -1346,8 +1380,27 @@ impl Component for EditorView {
         // clear with background color
         surface.set_style(area, cx.editor.theme.get("ui.background"));
         let config = cx.editor.config();
+
+        // check if bufferline should be rendered
+        use helix_view::editor::BufferLine;
+        let use_bufferline = match config.bufferline {
+            BufferLine::Always => true,
+            BufferLine::Multiple if cx.editor.documents().count() > 1 => true,
```
[the remark this answers is not in the document]
Comment (sudormrfbin) [reply helix-term/src/ui/editor.rs:1388]: `is_empty()` would return true even when there's only one doc, whereas we want it to be true only if there are >= 2 docs. So we would need something like:

```suggestion
            BufferLine::Multiple if cx.editor.documents().nth(1).is_some() => true, 
```

File: helix-term/src/ui/editor.rs [hunk 12]
```diff
@@ -578,6 +578,40 @@ impl EditorView {
...
+                editor
+                    .theme
+                    .try_get("ui.bufferline")
+                    .unwrap_or_else(|| editor.theme.get("ui.statusline"))
+            };
+
+            let text = format!(" {}{} ", fname, if doc.is_modified() { "[+]" } else { "" });
+            let offset = text.len();
+
+            surface.set_string(1 + viewport.x + len as u16, viewport.y, text, style);
```
[the remark this answers is not in the document]
Comment (aaron404) [reply helix-term/src/ui/editor.rs:616]: my understanding of `set_string` was that it automatically clipped the string to surface bounds. Is this not the case?

File: helix-term/src/ui/editor.rs [hunk 13]
```diff
@@ -1346,8 +1386,27 @@ impl Component for EditorView {
         // clear with background color
         surface.set_style(area, cx.editor.theme.get("ui.background"));
         let config = cx.editor.config();
+
+        // check if bufferline should be rendered
+        use helix_view::editor::BufferLine;
+        let use_bufferline = match config.bufferline {
+            BufferLine::Always => true,
+            BufferLine::Multiple if cx.editor.documents().count() > 1 => true,
```
Comment (aaron404) [inline helix-term/src/ui/editor.rs:1394]: What about this, just grab the size of the BTree directly?
```suggestion
            BufferLine::Always => true,
            BufferLine::Multiple if cx.editor.documents.len() > 1 => true,
```

File: book/src/configuration.md (hunk 5 above)
Comment (aaron404) [reply book/src/configuration.md:48]: I'll go with Never by default, based on @archseer's comment below

File: helix-term/src/ui/editor.rs [hunk 14]
```diff
@@ -1346,8 +1380,27 @@ impl Component for EditorView {
...
+            _ => false,
+        };
+
+        // -1 for commandline and -1 for bufferline
+        let editor_area = if use_bufferline {
+            area.clip_bottom(1).clip_top(1)
+        } else {
+            area.clip_bottom(1)
```
[the remark this answers is not in the document]
Comment (aaron404) [reply helix-term/src/ui/editor.rs:1402]: is it ok to make `editor_area` `mut` then? I generally avoid mut unless it's required. If this is the preference I'll change it though

File: helix-term/src/ui/editor.rs [hunk 15]
```diff
@@ -578,6 +578,50 @@ impl EditorView {
         }
     }
 
+    /// Render bufferline at the top
+    pub fn render_bufferline(editor: &Editor, viewport: Rect, surface: &mut Surface) {
+        let editor = PathBuf::from(SCRATCH_BUFFER_NAME); // default filename to use for scratch buffer
```
Comment (aaron404) [inline helix-term/src/ui/editor.rs:583]:
```suggestion
        let scratch = PathBuf::from(SCRATCH_BUFFER_NAME); // default filename to use for scratch buffer
```

File: helix-term/src/ui/editor.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (archseer) [reply helix-term/src/ui/editor.rs:1406]: Yeah it's fine

File: helix-term/src/ui/editor.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (archseer) [reply helix-term/src/ui/editor.rs:616]: It doesn't no, it uses `usize::MAX`. Any rendering functions will panic if they go offscreen.

**Comment by aaron404** [comment]
I am not sure what is holding this up. Github says there are changes requested but it looks like all the conversations are resolved.

**Comment by the-mikedavis** [comment]
Sorry about the wait - this is just waiting on the next round of review. It can take a while since there are so many PRs in the queue

File: helix-term/src/ui/editor.rs [hunk 16]
```diff
@@ -582,6 +582,57 @@ impl EditorView {
...
+                text,
+                surface.area.width as usize,
+                style,
+            );
+
+            len += offset;
+
+            if len > surface.area.width as usize {
+                break;
+            }
```
Comment (sudormrfbin) [inline helix-term/src/ui/editor.rs:632]: The offset calculation for `len` will have to use `text.width()` instead of `text.len()`, but we can simply use the return value of `surface.set_string()` (a tuple `(x, y)` with the end position) which does all of the offset calculation for us:

```suggestion
            let used_width = viewport.x - x; // x is initialized to viewport.x before the loop
            let rem_width = surface.area.width.saturating_sub(used_width);
            x = surface
                .set_stringn(x, viewport.y, text, rem_width as usize, style)
                .0;

            if x >= surface.area.right() {
                break;
            }
```

A better solution would be to have a version of [`Buffer::set_spans()`](https://github.com/helix-editor/helix/blob/a766b32ed197221aca7a9fec3af805b4d04a6f1d/helix-tui/src/buffer.rs#L418) that takes an iterator of `Span`s (a Span is a string + a style). Then the width and offset calculation would be managed for us if we output the buffer names from the iterator. But a plain for loop seems good enough for now 👍🏾.

File: helix-term/src/ui/editor.rs [hunk 17]
```diff
@@ -582,6 +582,57 @@ impl EditorView {
...
+            let style = if view!(editor).doc == doc.id() {
+                editor
+                    .theme
+                    .try_get("ui.bufferline.active")
+                    .unwrap_or_else(|| editor.theme.get("ui.background"))
+            } else {
+                editor
+                    .theme
+                    .try_get("ui.bufferline")
+                    .unwrap_or_else(|| editor.theme.get("ui.statusline"))
```
Comment (sudormrfbin) [inline helix-term/src/ui/editor.rs:614]: The theme style values should be calculated once outside the loop and reused on every iteration.

**Comment by sudormrfbin** [comment]
CI seems to be failing on [`SCRATCH_BUFFER_NAME` not in scope][1], but it indeed is and builds locally too.

[1]: https://github.com/helix-editor/helix/runs/7492588246?check_suite_focus=true#step:6:21

**Comment by aaron404** [comment]
> CI seems to be failing on [`SCRATCH_BUFFER_NAME` not in scope](https://github.com/helix-editor/helix/runs/7492588246?check_suite_focus=true#step:6:21), but it indeed is and builds locally too.

I merged master into it, maybe that will help

**Comment by sbromberger** [comment]
Any chance this might be reviewed again (and hopefully merged) soon? I've been using it for about a month and have come to depend on it. Would be nice to have one fewer thing that requires a custom build. Thanks!

**Comment by cor** [comment]
This looks fantastic! Love the amount of context it provides. I hope it gets merged soon

File: helix-term/src/ui/editor.rs [hunk 18]
```diff
@@ -592,6 +592,57 @@ impl EditorView {
...
+        for doc in editor.documents() {
+            let fname = doc
+                .path()
+                .unwrap_or(&scratch)
+                .file_name()
+                .unwrap_or_default()
+                .to_str()
+                .unwrap_or_default();
+
+            let style = if view!(editor).doc == doc.id() {
```
Comment (archseer) [inline helix-term/src/ui/editor.rs:626]: `let current_doc = view!(editor).doc` before the for loop to avoid fetching the current doc each time it loops

**Review by archseer — CHANGES_REQUESTED** [review]

File: helix-term/src/ui/editor.rs [hunk 19]
```diff
@@ -138,7 +138,7 @@ impl EditorView {
             highlights,
             &editor.config(),
         );
-        Self::render_gutter(editor, doc, view, view.area, surface, theme, is_focused);
+        Self::render_gutter(editor, doc, view, area, surface, theme, is_focused);
```
Comment (archseer) [inline helix-term/src/ui/editor.rs:141]: Why the change here?

File: helix-term/src/ui/editor.rs [hunk 20]
```diff
@@ -592,6 +592,57 @@ impl EditorView {
...
+            editor
+                .theme
+                .try_get("ui.bufferline.background")
+                .unwrap_or_else(|| editor.theme.get("ui.statusline")),
+        );
+
+        let bufferline_active = editor
+            .theme
+            .try_get("ui.bufferline.active")
+            .unwrap_or_else(|| editor.theme.get("ui.background"));
```
Comment (archseer) [inline helix-term/src/ui/editor.rs:609]: I think we could use `ui.statusline` and `ui.statusline.inactive` as fallback

File: helix-term/src/ui/editor.rs (hunk 19 above)
Comment (aaron404) [reply helix-term/src/ui/editor.rs]: I honestly cannot remember.. I have reverted it for now

**Review by archseer — APPROVED** [review]

**Comment by hemedani** [comment]
how can I active the bufferline ?

**Comment by CptPotato** [comment]
> how can I activate the bufferline ?

You can set `bufferline` to `always`, `never` or `multiple` in the editor section of your helix config, see https://docs.helix-editor.com/master/configuration.html#editor-section.

**Comment by hemedani** [comment]
tanks

**Comment by yashpalgoyal1304** [comment]
> > It could be nice to have an option that, similarly to some other editors, only shows the bufferline when there is more than one buffer to display in it.
> 
> That seems like good default behavior without needing a config option.

i came here to say exactly that, multiple should be default?!

**Comment by yashpalgoyal1304** [comment]
> Unchanged files with check annotations (Beta)
> Check warning on line 375 in helix-term/src/ui/picker.rs
>
> GitHub Actions / Check (msrv)
> use of deprecated associated function `retain_mut::RetainMut::retain_mut`: Rust 1.61 has included retain_mut directly


What does this mean??
